### ISSUE TITLE
Crash with "failed to move memtable to cache: std::bad_alloc"

### ISSUE THREAD

**Issue by locked** [scylladb/scylla#773]
Hello,

Scylla release 0.14.1-20160105.d3a0573 (ubuntu 14.04), is crashing at startup:

```
...
Seastar HTTP server listening on 0.0.0.0:10000 ...
Start Storage service ...
Exiting on unhandled exception of type 'std::bad_alloc': std::bad_alloc
```

The following message appear in /var/log/daemon.log:

```
Jan  8 09:58:19 fr-hq-casbkp-08 scylla[473]:  [shard 0] database - Populating Keyspace system_traces
Jan  8 09:58:19 fr-hq-casbkp-08 scylla[473]:  [shard 0] database - Keyspace /var/lib/scylla/data/system_traces: Reading CF events 
Jan  8 09:58:19 fr-hq-casbkp-08 scylla[473]:  [shard 0] database - Keyspace /var/lib/scylla/data/system_traces: Reading CF sessions 
Jan  8 09:58:19 fr-hq-casbkp-08 scylla[473]:  [shard 0] compaction_manager - compaction task handler stopped due to shutdown
Jan  8 09:58:19 fr-hq-casbkp-08 scylla[473]:  [shard 0] compaction_manager - compaction task handler stopped due to shutdown
Jan  8 09:58:19 fr-hq-casbkp-08 scylla[473]:  [shard 1] compaction_manager - compaction task handler stopped due to shutdown
Jan  8 09:58:19 fr-hq-casbkp-08 scylla[473]:  [shard 1] compaction_manager - compaction task handler stopped due to shutdown
Jan  8 09:58:19 fr-hq-casbkp-08 scylla[473]:  [shard 0] database - failed to move memtable to cache: std::bad_alloc (std::bad_alloc)
```

**Comment by penberg** [comment]
You say it's crashing at startup but the logs say that compaction manager is stopped because shutdown was issued. Did you stop Scylla yourself?

As for the exception, looks like ``row_cache::update()`` throws ``std::bad_alloc``. @avikivity @tgrabiec thoughts?

@syuu1228 How does one change default log level on Ubuntu?

**Comment by locked** [comment]
My understanding of the shutdown messages was that it was caused by the error, and it was happening before in the logs because of parallelism.

I attached the log file generated by launching scylla manually: scylla --developer-mode=1 > 2016-01-08_scylla.log 2>&1

[2016-01-08_scylla.log.gz](https://github.com/scylladb/scylla/files/82643/2016-01-08_scylla.log.gz)

The warning messages concerning "DateTieredCompactionStrategy" have always been there with previous releases and was never an issue.

**Comment by tgrabiec** [comment]
@locked How much memory and cores do you have in the system (`lstopo --of console --no-io`)?

**Comment by tgrabiec** [comment]
The `std::bad_alloc` which caused the failure of moving memtable to cache is not critical, it's not propagated, but it may be a symptom of running at low memory. The exception which kills the server is thrown from a different place.

@locked Is the database clean? If not, any chance you could send us your sstables?

**Comment by locked** [comment]
$ lstopo --of console --no-io
Machine (2938MB)
  Socket L#0 + L2 L#0 (4096KB) + L1d L#0 (32KB) + L1i L#0 (32KB) + Core L#0 + PU L#0 (P#0)
  Socket L#1 + L2 L#1 (4096KB) + L1d L#1 (32KB) + L1i L#1 (32KB) + Core L#1 + PU L#1 (P#1)

The database is clean since it starts fine with a build of commit b34a1f6a8 (December 16th).

Unfortunately I cannot send you our sstables since its big (300G+) and contains private data.

Etienne

**Comment by locked** [comment]
Ok it starts with 5G of RAM, is it expected to crash with 3G of RAM ? it seems a reasonable amount, since we dont expect any performance from the cluster (it's just a restore cluster).
Maybe a better error message to help address the issue faster ?
Thank you,

**Comment by tgrabiec** [comment]
@locked Could you send the output of `nodetool cfstats`?

**Comment by locked** [comment]
I removed the tables names:

[2016-01-08_cfstats.txt.gz](https://github.com/scylladb/scylla/files/82768/2016-01-08_cfstats.txt.gz)

**Comment by penberg** [comment]
@locked I totally agree that the error message should be improved upon. @tgrabiec can you send a patch for that?

**Comment by tgrabiec** [comment]
@penberg Which message are you suggesting to improve exactly?

**Comment by penberg** [comment]
This error message:

```
Jan  8 09:58:19 fr-hq-casbkp-08 scylla[473]:  [shard 0] database - failed to move memtable to cache: std::bad_alloc (std::bad_alloc)
```

s/std::bad_alloc/out-of-memory/ would already make it more readable, I think.

**Comment by avikivity** [comment]
Largest partition is 43MB (on disk), so this should have worked on 1.5GB shards.

It could be that because we're starting up, all column families are compacted concurrently, plus the 4-mutation queues we use during compaction cause an overload.  Maybe our queues should be also limited on the amount of memory, there's no need to queue multiple multi-megabyte partitions.

It could also be due to large blobs.  @locked - what's the maximum blob size you use?

**Comment by tgrabiec** [comment]
@avikivity I can see that some tables have a lot of sstables (~60). During compaction many sstables are merged into one and each sstable reader will keep its latest partition in memory (for merge sort). This probably depends on compaction strategy in use and sstable situation, but the number of large partitions loaded into memory can be much larger than 2 per shard.

**Comment by slivne** [comment]
I think any thresholds or checks done before the read starts won't alone prevent oom in general, given high enough demand from concurrent readers or other operations, we may run out of memory in the middle of the read, or despite even low thresholds.

We can't just wait until we get std::bad_alloc because the std::bad_alloc can be thrown not in the context of the read causing it but anywhere else, in the worst case in future::then_wrapped() triggering an abort. We should strive to make code paths exception safe as much as possible, but even then, an over-large read should fail itself and not some innocent operation.

Failing the most obtrusive reader would require ability to fail readers and reclaim their memory asynchronously, when they are off cpu. That's doable but complicated. We could start with cooperative failing, which is much easier. We would fail readers in their context when the amount of free memory in the system (with cache counting as free memory) falls below a threshold, or when the current reader allocates too much. Downside of cooperative failure is that we may need to fail not the most obtrusive reader, but the one which happens to hit the limit first.

We can achieve that by running reader in the context of an allocation_strategy (as in allocation_strategy.hh) which will do the memory accounting and fail allocation when we desire. The mutation code is already using allocation_strategy to allocate data, so we should automatically see the footprint of any temporary mutations created by readers.

The allocation_strategy would have to be set at the top-most level, which manages both the reader and mutations returned by it. There are few places like this:
  - column_family::query()
  - checksum_range_shard()
  - shard_reader

**Comment by gleb-cloudius** [comment]
On Wed, Mar 09, 2016 at 01:26:03AM -0800, slivne wrote:
> 
> I think any thresholds or checks done before the read starts won't alone prevent oom in general, given high enough demand from concurrent readers or other operations, we may run out of memory in the middle of the read, or despite even low thresholds.
> 
We do limit concurrent CQL readers/RPC readers by how much memory they
consume and do not allow new request to be send while memory does not
go below the threshold.

> We can't just wait until we get std::bad_alloc because the std::bad_alloc can be thrown not in the context of the read causing it but anywhere else, in the worst case in future::then_wrapped() triggering an abort. We should strive to make code paths exception safe as much as possible, but even then, an over-large read should fail itself and not some innocent operation.
> 
> Failing the most obtrusive reader would require ability to fail readers and reclaim their memory asynchronously, when they are off cpu. That's doable but complicated. We could start with cooperative failing, which is much easier. We would fail readers in their context when the amount of free memory in the system (with cache counting as free memory) falls below a threshold, or when the current reader allocates too much. Downside of cooperative failure is that we may need to fail not the most obtrusive reader, but the one which happens to hit the limit first.
> 
> We can achieve that by running reader in the context of an allocation_strategy (as in allocation_strategy.hh) which will do the memory accounting and fail allocation when we desire. The mutation code is already using allocation_strategy to allocate data, so we should automatically see the footprint of any temporary mutations created by readers.
> 
> The allocation_strategy would have to be set at the top-most level, which manages both the reader and mutations returned by it. There are few places like this:
>   - column_family::query()
>   - checksum_range_shard()
>   - shard_reader
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/scylladb/scylla/issues/773#issuecomment-194202797

--
			Gleb.

**Comment by tgrabiec** [comment]
2016-03-09 10:35 GMT+01:00 Gleb Natapov <notifications@github.com>:

> On Wed, Mar 09, 2016 at 01:26:03AM -0800, slivne wrote:
> >
> > I think any thresholds or checks done before the read starts won't alone
> prevent oom in general, given high enough demand from concurrent readers or
> other operations, we may run out of memory in the middle of the read, or
> despite even low thresholds.
> >
> We do limit concurrent CQL readers/RPC readers by how much memory they
> consume and do not allow new request to be send while memory does not
> go below the threshold.


Admission control happens when the request enters the system. We need this
as well, but in case of reads, it's hard to predict how much memory a read
will need given current design. So we may allow reads in even though they
won't be able to complete. Because of that, I think we need safety measures
which will fail reads before we run out of memory.

For the future, we could consider making readers stream data and thus their
footprint would be more predictable.

**Comment by gleb-cloudius** [comment]
On Thu, Mar 10, 2016 at 03:12:56AM -0800, Tomasz Grabiec wrote:
> 2016-03-09 10:35 GMT+01:00 Gleb Natapov <notifications@github.com>:
> 
> > On Wed, Mar 09, 2016 at 01:26:03AM -0800, slivne wrote:
> > >
> > > I think any thresholds or checks done before the read starts won't alone
> > prevent oom in general, given high enough demand from concurrent readers or
> > other operations, we may run out of memory in the middle of the read, or
> > despite even low thresholds.
> > >
> > We do limit concurrent CQL readers/RPC readers by how much memory they
> > consume and do not allow new request to be send while memory does not
> > go below the threshold.
> 
> 
> Admission control happens when the request enters the system. We need this
> as well, but in case of reads, it's hard to predict how much memory a read
> will need given current design. So we may allow reads in even though they
> won't be able to complete. Because of that, I think we need safety measures
> which will fail reads before we run out of memory.
> 
Of course. There is nothing we can do about a read request that needs
more memory that single shard has except streaming. But since we have
admission control db can provide feedback to cql server by serving reads,
which become to be big, serially. If all reads are like that eventually cql
server admission control will stop get new requests and thus preventing
ooming because of too big concurrency.

> For the future, we could consider making readers stream data and thus their
> footprint would be more predictable.
> 
> 
> >
> >
> > > We can't just wait until we get std::bad_alloc because the
> > std::bad_alloc can be thrown not in the context of the read causing it but
> > anywhere else, in the worst case in future::then_wrapped() triggering an
> > abort. We should strive to make code paths exception safe as much as
> > possible, but even then, an over-large read should fail itself and not some
> > innocent operation.
> > >
> > > Failing the most obtrusive reader would require ability to fail readers
> > and reclaim their memory asynchronously, when they are off cpu. That's
> > doable but complicated. We could start with cooperative failing, which is
> > much easier. We would fail readers in their context when the amount of free
> > memory in the system (with cache counting as free memory) falls below a
> > threshold, or when the current reader allocates too much. Downside of
> > cooperative failure is that we may need to fail not the most obtrusive
> > reader, but the one which happens to hit the limit first.
> > >
> > > We can achieve that by running reader in the context of an
> > allocation_strategy (as in allocation_strategy.hh) which will do the memory
> > accounting and fail allocation when we desire. The mutation code is already
> > using allocation_strategy to allocate data, so we should automatically see
> > the footprint of any temporary mutations created by readers.
> > >
> > > The allocation_strategy would have to be set at the top-most level,
> > which manages both the reader and mutations returned by it. There are few
> > places like this:
> > > - column_family::query()
> > > - checksum_range_shard()
> > > - shard_reader
> > >
> > > ---
> > > Reply to this email directly or view it on GitHub:
> > > https://github.com/scylladb/scylla/issues/773#issuecomment-194202797
> >
> > --
> > Gleb.
> >
> > �
> > Reply to this email directly or view it on GitHub
> > <https://github.com/scylladb/scylla/issues/773#issuecomment-194208810>.
> >
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/scylladb/scylla/issues/773#issuecomment-194795070

--
			Gleb.

**Comment by tgrabiec** [comment]
2016-03-10 12:21 GMT+01:00 Gleb Natapov <notifications@github.com>:

> On Thu, Mar 10, 2016 at 03:12:56AM -0800, Tomasz Grabiec wrote:
> > 2016-03-09 10:35 GMT+01:00 Gleb Natapov <notifications@github.com>:
> >
> > > On Wed, Mar 09, 2016 at 01:26:03AM -0800, slivne wrote:
> > > >
> > > > I think any thresholds or checks done before the read starts won't
> alone
> > > prevent oom in general, given high enough demand from concurrent
> readers or
> > > other operations, we may run out of memory in the middle of the read,
> or
> > > despite even low thresholds.
> > > >
> > > We do limit concurrent CQL readers/RPC readers by how much memory they
> > > consume and do not allow new request to be send while memory does not
> > > go below the threshold.
> >
> >
> > Admission control happens when the request enters the system. We need
> this
> > as well, but in case of reads, it's hard to predict how much memory a
> read
> > will need given current design. So we may allow reads in even though they
> > won't be able to complete. Because of that, I think we need safety
> measures
> > which will fail reads before we run out of memory.
> >
> Of course. There is nothing we can do about a read request that needs
> more memory that single shard has except streaming. But since we have
> admission control db can provide feedback to cql server by serving reads,
> which become to be big, serially. If all reads are like that eventually cql
> server admission control will stop get new requests and thus preventing
> ooming because of too big concurrency.


Can you describe this mechanism in more details, exactly when and how do we
signal? Since we do not know how much memory a read will need, we don't
know how much concurrency is too much before it happens. When it happens,
we could signal shard's admission control that we can't accept more, but
it's already too late to prevent oom.

**Comment by gleb-cloudius** [comment]
On Thu, Mar 10, 2016 at 03:44:03AM -0800, Tomasz Grabiec wrote:
> 2016-03-10 12:21 GMT+01:00 Gleb Natapov <notifications@github.com>:
> 
> > On Thu, Mar 10, 2016 at 03:12:56AM -0800, Tomasz Grabiec wrote:
> > > 2016-03-09 10:35 GMT+01:00 Gleb Natapov <notifications@github.com>:
> > >
> > > > On Wed, Mar 09, 2016 at 01:26:03AM -0800, slivne wrote:
> > > > >
> > > > > I think any thresholds or checks done before the read starts won't
> > alone
> > > > prevent oom in general, given high enough demand from concurrent
> > readers or
> > > > other operations, we may run out of memory in the middle of the read,
> > or
> > > > despite even low thresholds.
> > > > >
> > > > We do limit concurrent CQL readers/RPC readers by how much memory they
> > > > consume and do not allow new request to be send while memory does not
> > > > go below the threshold.
> > >
> > >
> > > Admission control happens when the request enters the system. We need
> > this
> > > as well, but in case of reads, it's hard to predict how much memory a
> > read
> > > will need given current design. So we may allow reads in even though they
> > > won't be able to complete. Because of that, I think we need safety
> > measures
> > > which will fail reads before we run out of memory.
> > >
> > Of course. There is nothing we can do about a read request that needs
> > more memory that single shard has except streaming. But since we have
> > admission control db can provide feedback to cql server by serving reads,
> > which become to be big, serially. If all reads are like that eventually cql
> > server admission control will stop get new requests and thus preventing
> > ooming because of too big concurrency.
> 
> 
> Can you describe this mechanism in more details, exactly when and how do we
> signal? Since we do not know how much memory a read will need, we don't
> know how much concurrency is too much before it happens. When it happens,
> we could signal shard's admission control that we can't accept more, but
> it's already too late to prevent oom.
> 
We do not signal anything directly from db to cql server. cql server
estimates how much each request "cost" for admission purpose. It mat be
entirely off, but this is not important, what is important is that at
some point when X requests will be in the system it will not allow more
requests. Now comes db part. DB servers all requests in parallel, but
after some requests pass certain thresholds they are serialized. For
instance if a threshold is 5% of memory then if we have 3 reads that
reached that threshold only one of them will proceed to be served, all
others will wait. If all X requests are like that then cql will stop
admitting more requests. Of course if X*5% is more than 100% then we
will oom much earlier then cql admission control will kick in, but this
can be solved by serving all request serially (in db) if memory is low.

--
			Gleb.

**Comment by pdziepak** [comment]
On 9 March 2016 at 09:25, slivne <notifications@github.com> wrote:
>
> I think any thresholds or checks done before the read starts won't alone
prevent oom in general, given high enough demand from concurrent readers or
other operations, we may run out of memory in the middle of the read, or
despite even low thresholds.
>
> We can't just wait until we get std::bad_alloc because the std::bad_alloc
can be thrown not in the context of the read causing it but anywhere else,
in the worst case in future::then_wrapped() triggering an abort. We should
strive to make code paths exception safe as much as possible, but even
then, an over-large read should fail itself and not some innocent operation.
>
> Failing the most obtrusive reader would require ability to fail readers
and reclaim their memory asynchronously, when they are off cpu. That's
doable but complicated. We could start with cooperative failing, which is
much easier. We would fail readers in their context when the amount of free
memory in the system (with cache counting as free memory) falls below a
threshold, or when the current reader allocates too much. Downside of
cooperative failure is that we may need to fail not the most obtrusive
reader, but the one which happens to hit the limit first.
>
> We can achieve that by running reader in the context of an
allocation_strategy (as in allocation_strategy.hh) which will do the memory
accounting and fail allocation when we desire. The mutation code is already
using allocation_strategy to allocate data, so we should automatically see
the footprint of any temporary mutations created by readers.
>
> The allocation_strategy would have to be set at the top-most level, which
manages both the reader and mutations returned by it. There are few places
like this:
>
> column_family::query()
> checksum_range_shard()
> shard_reader

Unfortunately, setting allocation strategy for readers is more tricky than
that. with_allocator() doesn't work with continuations – it just sets a
shard-local variable. We have continuations in our read paths (for
instance, mutation readers return future<mutation_opt>). So the solution is
either to individually set allocation strategy in each relevant
continuation which is going to be very messy and it will be easy to miss
something or to make with_allocator() preserve allocation strategy settings
in the continuation chain.

A way to do that is to introduce a state that is local to the continuation
chain and make pointer to the current allocation strategy be part of that
state. What would be needed is a pointer shard-local pointer to the current
continuation chain state like: thread_local
lw_shared_ptr<continuation_chain_state> current_state. then(),
then_wrapped() and all other functions that create a new continuation will
have to make sure that before execution of the continuation the pointer
'current_state' is set to the same value as it was when the continuation
was created (i.e. during the call to then(), etc.) In other words
continuations will inherit the state from their parents.
Continuation chains can fork so changes to the state will have to be done
with copy-on-write. For instance:

lw_shared_ptr<continuation_chain_state> get_new_state() {
    if (!current_state) { return make_shared<continuation_chain_state>(); }
    if (current_state.unique()) { return current_state; }
    return make_shared<continuation_chain_state>(*current_state);
}

with_allocator(allocation_strategy& as, Func&& func) {
    auto old_state = current_state;
    current_state = get_new_state();
    current_state->set_allocation_strategy(as);
    auto f = func();
    current_state = old_state;
    return f;
}

There is also a matter of decoupling continuation chain state which belongs
to seastar and allocation strategy which belongs to scylla.
This can be dealt with by making the state keep just a vector of chars and
requiring all is users to to register possible continuation chain local
variables in global constructors. If we require the them to be POD it will
be pretty straightforward. (And the whole mechanism will look a bit like
strange TLS).

**Comment by dorlaor** [comment]
@slivne what about this?

**Comment by slivne** [comment]
closing this issue - since then we have limited the concurrent sstable readers.
